### PR TITLE
HTTP imports #2

### LINKED ISSUE #416: import via http
start here https://github.com/ry/deno/blob/6b49944da1e7b06aeba04b1c0ded8e3fdb3c436b/src/handlers.rs#L277

prototype implementation here: https://github.com/ry/deno/blob/golang/os.go#L123

### PR DESCRIPTION
**Pull request by ry**
This is mostly @NewLunarFire's work. I've rebased and made clean ups.
Continued from #519.

Depends on #518

Fixes #416

### CODE REVIEW

File: src/net.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  let url = module_name.parse::<Uri>().unwrap();
+  let client = Client::new();
+
+  // TODO Use Deno's RT
+  let mut rt = Runtime::new().unwrap();
+
+  let body = rt.block_on(
+    client
+      .get(url)
+      .and_then(move |response| response.into_body().concat2()),
```
Comment (NewLunarFire) [inline src/net.rs]: Do we need the move here? We're not borrowing anything.

Comment (ry) [reply src/net.rs]: Thanks - fixed.

File: src/deno_dir.rs [hunk 2]
```diff
@@ -395,24 +487,6 @@ fn test_resolve_module() {
...
-    unimplemented!();
-  } else {
-    assert!(
-      module_name == filename,
-      "if a module isn't remote, it should have the same filename"
-    );
-    fs::read_file_sync_string(Path::new(filename))
-  }
+fn is_remote(module_name: &str) -> bool {
+  module_name.starts_with("http")
```
[the remark this answers is not in the document]
Comment (NewLunarFire) [reply src/deno_dir.rs]: @ry We should replace this with module_name.starts_with("http://") to make sure we don't match against local file names that start with http (e.g. import "http-parser.ts")

[the remark this answers is not in the document]
Comment (ry) [reply src/deno_dir.rs]: Sure. I'll leave that as something to be fixed with a test. (I expect this code to be rewritten soon.)

File: src/deno_dir.rs [hunk 3]
```diff
@@ -189,27 +248,59 @@ impl DenoDir {
...
-
+    debug!("module_name: {}, filename: {}", module_name, filename);
     Ok((module_name, filename))
   }
 }
 
+fn get_cache_filename(basedir: &Path, url: Url) -> PathBuf {
+  let mut out = basedir.to_path_buf();
+  out.push(url.host_str().unwrap());
+  for path_seg in url.path_segments().unwrap() {
```
Comment (piscisaureus) [inline src/deno_dir.rs]: I have a hunch that we're gonna find a security hole or two in this code.
Maybe it'd be better to use `sha(path)` as the cache filename.

File: src/deno_dir.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (NewLunarFire) [reply src/deno_dir.rs]: How is that for a test and an implementation?

```rust
extern crate url;

use url::Url;

fn main() {
    assert_eq!(is_remote("http-import.ts"), false);
    assert_eq!(is_remote("http/file.ts"), false);
    assert_eq!(is_remote("/home/tommy/http.ts"), false);
    assert_eq!(is_remote("C:\\Users\\tommy\\http.ts"), false);
    assert_eq!(is_remote("http://example.com/http.ts"), true);
    assert_eq!(is_remote("https://example.com/http.ts"), true);
}

fn is_remote(module_name: &str) -> bool {
    let url = Url::parse(module_name);
    if url.is_err() {
        return false;
    }

    match url.unwrap().scheme() {
        "http" => true,
        "https" => true,
        _ => return false
    }
}
```

**Comment by ry** [comment]
@piscisaureus I don't disagree that there could be better organization - but cache get/set have proper interfaces defined as flatbuffer messages:
https://github.com/denoland/deno/blob/ecb955929fd528f19c5fab9e4e904b621d8cf94d/src/msg.fbs#L83-L102

**Pull request merged**